File: 142-LinkedListCycleIi/142-LinkedListCycleIi.py
# Last updated: 9/14/2025, 9:16:33 AM
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        node=head
        visited=set()
        while(node is not None):
            if node in visited:
                return node
            visited.add(node)
            node=node.next
        return None

File: 142-LinkedListCycleIi/142-LinkedListCycleIi.java
// Last updated: 9/14/2025, 9:16:29 AM
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        ListNode fast = head;
        ListNode slow = head;

        while(fast!=null && fast.next!=null){
            fast = fast.next.next;
            slow = slow.next;
            if(fast == slow){
                slow = head;
                while(fast != slow){
                    fast = fast.next;
                    slow = slow.next;
                }
                return slow;
            }
        }
        return null;
    }
}